---
title: "Linux Command Line"
author: "Rachel Wood"
date: "2023-03-22"
output:
  html_document:
    highlight: espresso
  pdf_document:
    highlight: espresso
header-includes:
- \usepackage{tikz}
- \usepackage{fvextra}
- \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
---

```{r setup, include=FALSE}
library(formatR)
library(knitr)
library(highr)
opts_chunk$set(eval = FALSE, engine = 'bash')
hook_output = knit_hooks$get('source',)
opts_knit$set(dev = 'pdf',
              pdf.options(encoding = "ISOLatin9.enc"),
              message = FALSE,
              warning = FALSE,
              tidy.opts = list(width.cutoff=50), tidy=TRUE,
              fig.pos ="H", out.width = '60%',
              fig.align = 'center',
              root.dir = "/home/ac18826/Documents/COMPASS/First_Year/TB2/Statistical_Computing/Intro_to_Linux/Portfolio/"
              )
```

For this portfolio I will use the example of bioinformatics to showcase the utility of the Linux command line. The data we will be using is the genome and protein datasets from the SARS-CoV-2 virus, which can be found at \url{https://www.ncbi.nlm.nih.gov/genome/?term=SARS-CoV-2}.

# Working with Files and File Systems

```{r, engine='bash', include = FALSE}
rm -r bioinf_ex
cd ..
cp GCF_009858895.2_ASM985889v3_genomic.fna.gz Portfolio/
cp GCF_009858895.2_ASM985889v3_protein.faa.gz Portfolio/

```


We start in a folder containing the files downloaded from the link above:

```{r, engine='bash', eval = FALSE}
ls
GCF_009858895.2_ASM985889v3_genomic.fna.gz  
GCF_009858895.2_ASM985889v3_protein.faa.gz
```

We first create a folder to work within using the `mkdir` command and use the `cd` command to move into this new file. Finally we can use the `ls` command to view the contents of this folder:
```{r, engine='bash'}
mkdir bioinf_ex

```

As expected, the directory is empty, but we can move the genome and protein files to our new folders. We use the \texttt{cd ..} to go one `step back' in the directory and return to our original directory. We use the \texttt{ls} command again and then the \texttt{mv} command to move our two files into the new folder we have created:




```{r, engine='bash'}
mv GCF_009858895.2_ASM985889v3_protein.faa.gz -t bioinf_ex/
mv GCF_009858895.2_ASM985889v3_genomic.fna.gz -t bioinf_ex/
cd bioinf_ex/
```
```{r, include=FALSE}
opts_knit$set(root.dir ="/home/ac18826/Documents/COMPASS/First_Year/TB2/Statistical_Computing/Intro_to_Linux/Portfolio/bioinf_ex/")
```
We now check these files have been successfully moved:
```{r, engine ='bash', eval = FALSE}
ls 
GCF_009858895.2_ASM985889v3_genomic.fna.gz  
GCF_009858895.2_ASM985889v3_protein.faa.gz
```


We notice the `.gz` file extension, meaning we need to unzip the files. We can do this with the `gzip` command:

```{r, engine = 'bash'}
gzip -d GCF_009858895.2_ASM985889v3_genomic.fna.gz
gzip -d GCF_009858895.2_ASM985889v3_protein.faa.gz
```

```{r, engine='bash', eval = FALSE}
ls
GCF_009858895.2_ASM985889v3_genomic.fna
GCF_009858895.2_ASM985889v3_protein.fa
```

We finally rename the files to something more convenient with the `mv` command:

```{bash}
mv GCF_009858895.2_ASM985889v3_genomic.fna genomic.fna
mv GCF_009858895.2_ASM985889v3_protein.faa protein.faa
```

# Genomic Data

This section focuses on the genomic.fna file. We first use the `head` command to view the first few lines of the file:
```{bash}
head genomic.fna
NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome
ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAA
AATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGG
ACACGAGTAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTGTTGCAGCCGATCATCAGCACATCTAGGTTT
CGTCCGGGTGTGACCGAAAGGTAAGATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGC
CTGTTTTACAGGTTCGCGACGTGCTCGTACGTGGCTTTGGAGACTCCGTGGAGGAGGTCTTATCAGAGGCACGTCAACAT
CTTAAAGATGGCACTTGTGGCTTAGTAGAAGTTGAAAAAGGCGTTTTGCCTCAACTTGAACAGCCCTATGTGTTCATCAA
ACGTTCGGATGCTCGAACTGCACCTCATGGTCATGTTATGGTTGAGCTGGTAGCAGAACTCGAAGGCATTCAGTACGGTC
GTAGTGGTGAGACACTTGGTGTCCTTGTCCCTCATGTGGGCGAAATACCAGTGGCTTACCGCAAGGTTCTTCTTCGTAAG
AACGGTAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCATTTGACTTAGGCGACGAGCTTGGCACTGA

```

We might want to know how long a genome is, this is often measured by the number of base pairs, here each pair is represented by a character - a base pair is either \texttt{A-T} or \texttt{G-C}. We use the \texttt{grep} command using the \texttt{-o} flag instead combined with the \texttt{wc} command to get the count of \texttt{A}, \texttt{C}, \texttt{G} or \texttt{T} characters in the file (excluding the header line):
```{bash}
tail -n+2 genomic.fna | grep -o [ATCG] | wc -l
29903
```
The square brackets indicate that we want want the count of any of these letters instead of the pattern `ATGC`

We can use similar commands to obtain the individual counts of the letters:

```{bash}
tail -n+2 genomic.fna | grep -o A | wc -l
8954
tail -n+2 genomic.fna | grep -o T | wc -l
9594
tail -n+2 genomic.fna | grep -o C | wc -l
5492
tail -n+2 genomic.fna | grep -o G | wc -l
5863
```

From this we can see the genome is \texttt{A-T} enriched. 

The start codon Methionine is represented by \texttt{ATG}, we can see how many times this appears. 

For this we first need to remove the header line and all new lines  characters (denoted `\n`) and save the changes in a new file:
```{bash}
tail -n+2 genomic.fna > genomic_new.fna
echo -n $(tr -d "\n" < genomic_new.fna) > genomic_new.fna 
```

Note the use of `echo`, this creates a subshell, which allows us to read from and write to the same file.

We can now search for the Methionine codon, using the `grep` command again,
```{bash}
grep -o ATG genomic_new.fna | wc -l
725
```

The covid nucleocapsid phosphoprotein starts with the motif ATGTCTGATAAT and ends with AACTCAGGCCTA. We want to isolate this gene, so find the character position of the character before the start of the patterns using `grep -b` (the `-b` indicates we want the byte off-set)
```{bash}
grep -aob ATGTCTGATAAT genomic_new.fna 
28273:ATGTCTGATAAT

grep -ob AACTCAGGCCTA genomic_new.fna 
29520:AACTCAGGCCTA
```

We can now write the gene to a new file 
